tests/parser/syntax/test_msg_data.py

```diff
@@ -159,3 +182,56 @@ def foo(_value: uint256) -> uint256:
 
     with pytest.raises(TransactionFailed):
         contract.foo(42)
+
+
+@pytest.mark.parametrize("offset", range(5))
+def test_extract32_at_the_bounds_of_calldata(get_contract, offset):
+    code = f"""
+@external
+def foo() -> bytes32:
+    value: bytes32 = extract32(msg.data, {offset})
+    return value
+"""
+    contract = get_contract(code)
+
+    if offset == 4:
+        # if offset == 4, we are do msg.data[4:36], which is just empty
+        # null bytes, we only allow users to extract up to but not including
+        # the bounds of calldata, so msg.data[3:35] is valid as it still includes
+        # the last bytes of the 4 byte function selector.
+        # If a user needs empty bytes, they can use the EMPTY_BYTES32 constant
+        with pytest.raises(TransactionFailed):
+            contract.foo()
+    else:
+        contract.foo()
+
+
+def test_more_calldata_than_expected(get_contract):
+    # this contract purposefully reads past the expected length of calldata
+    code = """
+@external
+def foo() -> uint256:
+    return extract32(msg.data, 4, output_type=uint256)
+"""
+    contract = get_contract(code)
+
+    with pytest.raises(TransactionFailed):
+        contract.foo()
+
+    # however, if we append data this will be fine
+    code_1 = """
+@external
+def bar(_addr: address, _num: uint256) -> uint256:
+    result: Bytes[32] = raw_call(
+        _addr,
+        concat(
```

Comment:
This is an interesting test but I'm not sure we should rely on this behavior. If vyper were stricter it might reject any calldata that is longer than expected.